File: Homework/Hoang_Nguyen/lesson7/508_mostFrequentSubtreeSum.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */
var findFrequentTreeSum = function (root) {
  // sum => freq
  const freqMap = new Map();
  let maxFreq = 0;

  function sumSubtree(node) {
    if (!node) return 0;
    const sum = sumSubtree(node.left) + sumSubtree(node.right) + node.val;
    const freq = (freqMap.get(sum) || 0) + 1;
    if (freq > maxFreq) {
      maxFreq = freq;
    }
    freqMap.set(sum, freq);
    return sum;
  }

  sumSubtree(root);

  let result = [];
  for (const [sum, freq] of freqMap) {
    if (freq === maxFreq) {
      result.push(sum);
    }
  }
  return result;
};

File: Homework/Ngan_Luu/508-most-frequent-subtree-sum.py
# 508. https://leetcode.com/problems/most-frequent-subtree-sum/
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def findFrequentTreeSum(self, root: Optional[TreeNode]) -> List[int]:
        freq={}
        def dfs(node):
            if not(node):
                return 0
            else:
                left_val=dfs(node.left)
                right_val=dfs(node.right)
                sum_node=left_val+right_val+node.val
                if sum_node in freq:
                    freq[sum_node]+=[node.val]
                else:
                    freq[sum_node]=[node.val]
                return sum_node
        dfs(root)
        max_freq=0
        for key,value in freq.items():
            if len(value)>max_freq:
                output=[key]
                max_freq=len(value)
            elif len(value)==max_freq:
                output.append(key)
        return output